File: android/app/src/main/java/com/tartunlp/neurokone/FastSpeechModel.java
package com.tartunlp.neurokone;

import android.annotation.SuppressLint;
import android.util.Log;

import org.tensorflow.lite.DataType;
import org.tensorflow.lite.Interpreter;
import org.tensorflow.lite.support.tensorbuffer.TensorBuffer;

import java.io.File;
import java.nio.FloatBuffer;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class FastSpeechModel {
    private final String TAG = "FastSpeech2";
    private final Interpreter model;

    private int voice = 0;
    private float speed = 1.0f; //0.1 - 6
    private float pitch = 1.0f; //0.25 - 4
    private float energy = 1.0f;

    public FastSpeechModel(String path) {
        this.model = new Interpreter(new File(path));
    }

    public void setVoice(int voice) {
        this.voice = voice;
    }

    public void setSpeed(float speed) {
        this.speed = 1.0f / speed;  //speed input is actually the audio length multiplier so the faster the speech, the lower value the input
    }

    public void setPitch(float pitch) { this.pitch = pitch; }

    public void setEnergy(float energy) { this.energy = energy; }

    public synchronized TensorBuffer getMelSpectrogram(int[] inputIds) {
        Log.d(TAG, "input id length: " + inputIds.length);
        this.model.resizeInput(0, new int[]{1, inputIds.length});
        this.model.allocateTensors();

        @SuppressLint("UseSparseArrays")
        Map<Integer, Object> outputMap = new HashMap<>();
        FloatBuffer outputBuffer = FloatBuffer.allocate(300000);
        outputMap.put(0, outputBuffer);

        int[][] inputs = new int[1][inputIds.length];
        inputs[0] = inputIds;

        Object[] input = new Object[]{inputs, new int[]{voice}, new float[]{speed}, new float[]{pitch}, new float[]{energy}};
        this.model.runForMultipleInputsOutputs(input, outputMap);

        int[] shape = this.model.getOutputTensor(0).shape();
        Log.d(TAG, "Spectrogram shape: " + Arrays.toString(shape));

        int[] newShape = {1, outputBuffer.position() / shape[2], shape[2]};
        TensorBuffer spectrogram = TensorBuffer.createFixedSize(newShape, DataType.FLOAT32);
        float[] outputArray = new float[outputBuffer.position()];
        outputBuffer.rewind();
        outputBuffer.get(outputArray);
        spectrogram.loadArray(outputArray);
        return spectrogram;
    }
}

File: macos/EestiTtsMacOS/FastSpeechModel.swift
//
//  FastSpeechModel.swift
//  EestiTtsMacOS
//
//  Created by Rasmus Lellep on 06.03.2024.
//

import os
import Foundation
import TensorFlowLite

class FastSpeechModel: TfLiteModel {
    private var voice: Int = 0;
    private var speed: Float = 1.0; //0.1 - 6
    private var pitch: Float = 1.0; //0.25 - 4
    private var energy: Float = 1.0;
    
    func setVoice(voice: Int) {
        self.voice = voice;
    }

    func setSpeed(speed: Float) {
        self.speed = 1.0 / speed;  //speed input is actually the audio length multiplier so the faster the speech, the lower value the input
    }

    func setPitch(pitch: Float) { self.pitch = pitch; }

    func setEnergy(energy: Float) { self.energy = energy; }
    
    func getMelSpectrogram(inputIds: [Int]) throws -> Data {
        let primaryInputs: [[Int32]] = [inputIds.map { Int32($0) }, [Int32(self.voice)]]
        let secondaryInputs: [[Float]] = [[self.speed], [self.pitch], [self.energy]]
        
        try self.model!.resizeInput(at: 0, to: Tensor.Shape([1, inputIds.count]))
        for id in 1..<primaryInputs.count {
            try self.model!.resizeInput(at: id, to: Tensor.Shape([primaryInputs[id].count]))
        }
        for id in 0..<secondaryInputs.count {
            try self.model!.resizeInput(at: id+primaryInputs.count, to: Tensor.Shape([secondaryInputs[id].count]))
        }
        
        try self.model!.allocateTensors()
        
        for id in 0..<primaryInputs.count {
            let dataIn = primaryInputs[id].withUnsafeBufferPointer(Data.init)
            try self.model!.copy(dataIn, toInputAt: id)
        }
        for id in 0..<secondaryInputs.count {
            let dataIn = secondaryInputs[id].withUnsafeBufferPointer(Data.init)
            try self.model!.copy(dataIn, toInputAt: id+primaryInputs.count)
        }
        
        // inference
        try self.model!.invoke()
        
        return try self.model!.output(at: 0).data
    }
}